### PR TITLE
`azurerm_kubernetes_flux_configuration` - added postBuild and wait properties to kustomization block

### LINKED ISSUE #24081: Support for postBuild substitution in flux_configuration
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Community Note

* Please vote on this issue by adding a :thumbsup: [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment and review the [contribution guide](https://github.com/hashicorp/terraform-provider-azurerm/blob/main/contributing/README.md) to help.

### Description

I would be great to have support to add postBuild substitutions in the azurerm_flux_configuration resource. Presently, this can be done using a patch request to the API but this is not ideal. The setting is also not yet exposed in the azure cli or azure portal.

https://registry.terraform.io/providers/hashicorp/azurerm/latest/docs/resources/kubernetes_flux_configuration#kustomizations

This is the api endpoint that can be used https://learn.microsoft.com/en-us/rest/api/kubernetesconfiguration/flux-configurations/update?view=rest-kubernetesconfiguration-2023-05-01&tabs=HTTP#kustomizationpatchdefinition

Here is an example request:
```
az rest --method patch --headers "{'Content-Type': 'application/json'}" --uri https://management.azure.com/subscriptions/<sub-id>/resourceGroups/<my-rg>/providers/Microsoft.ContainerService/managedClusters/<my-cluster>/providers/Microsoft.KubernetesConfiguration/fluxConfigurations/flux-system\?api-version\=2023-05-01 --body "{'properties':{'gitRepository':{'url':'https://github.com/my-flux-config.git','repositoryRef':{'branch':'my-branch'}},'kustomizations':{'flux-system':{'path':'./clusters/sbox/01','syncIntervalInSeconds':300,'postBuild':{'substitute':{'WI_NAME':'flux-system','ENVIRONMENT':'sbox','WI_ENVIRONMENT':'sbox','CLUSTER':'01','ISSUER_URL':'https://uksouth.oic.prod-aks.azure.com/<tenant-id>/<federated-credential-id>/','ENV_MONITOR_CHANNEL':'aks-monitor-sbox','KEYVAULT_ENVIRONMENT':'sbox'}}}}}}"
```

### New or Affected Resource(s)/Data Source(s)

azurerm_kubernetes_flux_configuration

### Potential Terraform Configuration

```hcl
resource "azurerm_kubernetes_flux_configuration" "microsoft_flux_configuration" {
  name       = "flux-system"
  cluster_id = azurerm_kubernetes_cluster.kubernetes_cluster.id
  namespace  = "flux-system"

  git_repository {
    url             = "https://github.com/my-flux-config"
    reference_type  = "branch"
    reference_value = "my-branch"
  }

  kustomizations {
    name = "flux-system"
    path = "./clusters/${var.environment}/${var.cluster_number}"
    postBuild = {
     substitute = {
      WI_NAME: "flux-system"
      ENVIRONMENT: "sbox"
      WI_ENVIRONMENT: "sbox"
      CLUSTER: "01"
      ISSUER_URL: "https://uksouth.oic.prod-aks.azure.com/<tenant-id>/<federated-credential-id>/"
      ENV_MONITOR_CHANNEL: "aks-monitor-sbox"
      KEYVAULT_ENVIRONMENT: "sbox"
      }
    }
  }

  scope = "cluster"

  depends_on = [
    azurerm_kubernetes_cluster_extension.microsoft_flux_extension
  ]
}
```

### References

_No response_

### PR DESCRIPTION
**Pull request by jayctran**
## Community Note

* Please vote on this PR by adding a :thumbsup: [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original PR to help the community and maintainers prioritize for review
* Please do not leave "+1" or "me too" comments, they generate extra noise for PR followers and do not help prioritize for review

## PR Checklist

- [X] I have followed the guidelines in our [Contributing Documentation](../blob/main/contributing/README.md).
- [X] I have checked to ensure there aren't other open [Pull Requests](../pulls) for the same update/change.
- [X] I have checked if my changes close any open issues. If so please include appropriate [closing keywords](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword) below.
- [X] I have updated/added Documentation as required written in a helpful and kind way to assist users that may be unfamiliar with the resource / data source.
- [X] I have used a meaningful PR title to help maintainers and other users understand this change and help prevent duplicate work. 
For example: “`resource_name_here` - description of change e.g. adding property `new_property_name_here`”

## Changes to existing Resource / Data Source

- [X] I have added an explanation of what my changes do and why I'd like you to include them (This may be covered by linking to an issue above, but may benefit from additional explanation).
- [X] I have written new tests for my resource or datasource changes & updated any relevent documentation.
- [ ] I have successfully run tests with my changes locally. If not, please provide details on testing challenges that prevented you running the tests.
- [ ] (For changes that include a state migration only). I have manually tested the migration path between relevant versions of the provider.

## Testing 

- [X] My submission includes Test coverage as described in the [Contribution Guide](../blob/main/contributing/topics/guide-new-resource.md) and the tests pass. (if this is not possible for any reason, please include details of why you did or could not add test coverage)

Every time I tried to run the acceptance tests I hit this issue and couldn't get past it.
image
image

Got this working in the project I needed this for:
image

## Change Log

Below please provide what should go into the changelog (if anything) conforming to the [Changelog Format documented here](../blob/main/contributing/topics/maintainer-changelog.md).

* `azurerm_kubernetes_flux_configuration` - added postBuild and wait properties to kustomization block [GH-24081]

This is a (please select all that apply):

- [ ] Bug Fix
- [ ] New Feature (ie adding a service, resource, or data source)
- [X] Enhancement
- [ ] Breaking Change

## Related Issue(s)
Fixes #24081

> [!NOTE] 
> If this PR changes meaningfully during the course of review please update the title and description as required.

### CODE REVIEW

**Review by katbyte — CHANGES_REQUESTED** [review]
Test failure:
```
------- Stdout: -------
=== RUN   TestAccKubernetesFluxConfiguration_kustomizationPostBuild
=== PAUSE TestAccKubernetesFluxConfiguration_kustomizationPostBuild
=== CONT  TestAccKubernetesFluxConfiguration_kustomizationPostBuild
    testcase.go:113: Step 1/2 error: Error running pre-apply refresh: exit status 1
        
        Error: Missing key/value separator
        
          on terraform_plugin_test.tf line 80, in resource "azurerm_kubernetes_flux_configuration" "test":
          75:     postBuild = {
          76:       substitute = {
          77:         "key" = "value",
          78:         "key2" = "value2"
          79:       }
          80:       substituteFrom {
        
        Expected an equals sign ("=") to mark the beginning of the attribute value.
    testing_new.go:79: Error retrieving state, there may be dangling resources: exit status 1
        
        Error: Missing key/value separator
        
          on terraform_plugin_test.tf line 80, in resource "azurerm_kubernetes_flux_configuration" "test":
          75:     postBuild = {
          76:       substitute = {
          77:         "key" = "value",
          78:         "key2" = "value2"
          79:       }
          80:       substituteFrom {
        
        Expected an equals sign ("=") to mark the beginning of the attribute value.
--- FAIL: TestAccKubernetesFluxConfiguration_kustomizationPostBuild (2.04s)
FAIL

```

**Comment by jayctran** [comment]
Hi all, I've rebased main to resolve a merge conflict that it came up with, and also resolved make depscheck.

What I'm not sure of and would like some guidance around is how would you want the tests to be created when a client id, secret is required - as the run-gradually-deprecated checks are failing?

It's failing where os.GetEnv("ARM_CLIENT_ID") etc is utilised but is the expected way to do this to create a service principal with the azuread provider? In these particular tests they require a service principal which can be provided - or can already existing service principals be leveraged since the User Assigned Managed Identity does not have secrets that can be used?

Basically... What would be the best approach to leverage an application registration where I need to input the client id, tenant id and client secret as part of the resource test? As I'm only modifying these tests and didn't come up with them, I'd be happy with the easiest approach to resolving this.

**Comment by stephybun** [comment]
Hey @jayctran,

The expectation is to create an application and service principal using the `azuread` provider. Here is a snippet of what that looks like for other tests that were previously using `os.GetEnv("ARM_CLIENT_ID")`.

https://github.com/hashicorp/terraform-provider-azurerm/blob/c18bce7f284e6e98ef4db2bb19d2e55c4d687c27/internal/services/containers/kubernetes_cluster_network_resource_test.go#L2298-L2315

**Comment by jayctran** [comment]
> Hey @jayctran,
> 
> The expectation is to create an application and service principal using the `azuread` provider. Here is a snippet of what that looks like for other tests that were previously using `os.GetEnv("ARM_CLIENT_ID")`.
> 
> https://github.com/hashicorp/terraform-provider-azurerm/blob/c18bce7f284e6e98ef4db2bb19d2e55c4d687c27/internal/services/containers/kubernetes_cluster_network_resource_test.go#L2298-L2315

Thanks @stephybun, that makes sense, I'll work on converting those older tests. Cheers

File: internal/services/containers/kubernetes_flux_configuration_resource_test.go [hunk 1]
```diff
@@ -715,3 +753,43 @@ resource "azurerm_kubernetes_flux_configuration" "test" {
...
+
+  kustomizations {
+    name = "kustomization-1"
+    path = "./test/path"
+    postBuild {
+      substitute = {
+        "key"  = "value",
+        "key2" = "value2"
+      }
+      substituteFrom {
```
Comment (katbyte) [inline internal/services/containers/kubernetes_flux_configuration_resource_test.go:781]: I think this should be 
```suggestion
    post_build {
      substitute = {
        "key"  = "value",
        "key2" = "value2"
      }
      substitute_from {
```

as we have a test failure:
```
------- Stdout: -------
=== RUN   TestAccKubernetesFluxConfiguration_kustomizationPostBuild
=== PAUSE TestAccKubernetesFluxConfiguration_kustomizationPostBuild
=== CONT  TestAccKubernetesFluxConfiguration_kustomizationPostBuild
    testcase.go:113: Step 1/2 error: Error running pre-apply refresh: exit status 1
        
        Error: Unsupported block type
        
          on terraform_plugin_test.tf line 75, in resource "azurerm_kubernetes_flux_configuration" "test":
          75:     postBuild {
        
        Blocks of type "postBuild" are not expected here. Did you mean "post_build"?
--- FAIL: TestAccKubernetesFluxConfiguration_kustomizationPostBuild (3.25s)
FAIL
```

**Review by katbyte — CHANGES_REQUESTED** [review]

**Comment by jayctran** [comment]
Have just rebased this to get this up to date with main

File: internal/services/containers/kubernetes_flux_configuration_resource_test.go (hunk 1 above)
Comment (jayctran) [reply internal/services/containers/kubernetes_flux_configuration_resource_test.go]: Fixed this in latest commit

**Comment by jayctran** [comment]
hi @katbyte @stephybun , following up on whether the tests could be triggered again for this? The issues with the tests should be fixed now. Thanks

**Comment by jayctran** [comment]
@katbyte  bumping this

**Review by ms-zhenhua — CHANGES_REQUESTED** [review]
Hi @jayctran ,

Thanks for this PR - I've taken a look through and left some comments inline. If we can fix those up, this should be good to go 👍

File: website/docs/r/kubernetes_flux_configuration.html.markdown [hunk 2]
```diff
@@ -104,6 +114,10 @@ A `kustomizations` block supports the following:
 
 * `depends_on` - (Optional) Specifies other kustomizations that this kustomization depends on. This kustomization will not reconcile until all dependencies have completed their reconciliation.
 
+* `post_build` - (Optional) A `post_build` block as defined below.
+
+* `wait` - (Optional) Enable/disable health check for all Kubernetes objects created by this Kustomization.
```
[the remark this answers is not in the document]
Comment (jayctran) [reply website/docs/r/kubernetes_flux_configuration.html.markdown:119]: Thanks for the review - I've left it as "wait" as it's the property used for the FluxConfiguration API as well as what's documented for Flux kustomizations: https://fluxcd.io/flux/components/kustomize/kustomizations/#wait 

I figured using the official naming would enable any Flux user to understand and reference the official documentation.

File: website/docs/r/kubernetes_flux_configuration.html.markdown [hunk 3]
```diff
@@ -192,6 +206,22 @@ A `git_repository` block supports the following:
...
+
+A `post_build` block supports the following:
+* `substitute` - (Optional) Specifies the key/value pairs holding the variables to be substituted in this Kustomization.
+
+* `substitute_from` - (Optional) A `substitute_from` block as defined below.
+
+---
+
+A `substitute_from` block supports the following:
+* `kind` - (Required) Define whether it is ConfigMap or Secret that holds the variables to be used in substitution.
```
[the remark this answers is not in the document]
Comment (jayctran) [reply website/docs/r/kubernetes_flux_configuration.html.markdown:219]: That's a good idea, there's only two options for this - I will change this to an enum

File: internal/services/containers/kubernetes_flux_configuration_resource.go [hunk 4]
```diff
@@ -220,6 +233,52 @@ func (r KubernetesFluxConfigurationResource) Arguments() map[string]*pluginsdk.S
...
+												Optional: true,
+												Default:  true,
+											},
+										},
+									},
+								},
+							},
+						},
+					},
+					"wait": {
```
[the remark this answers is not in the document]
Comment (jayctran) [reply internal/services/containers/kubernetes_flux_configuration_resource.go:277]: I'll update the defaults with the same defaults in the api

File: website/docs/r/kubernetes_flux_configuration.html.markdown (hunk 2 above)
Comment (ms-zhenhua) [inline website/docs/r/kubernetes_flux_configuration.html.markdown:119]:
```suggestion
* `wait` - (Optional) Whether to enable health check for all Kubernetes objects created by this Kustomization. Defaults to `true`.
```

File: website/docs/r/kubernetes_flux_configuration.html.markdown (hunk 3 above)
Comment (ms-zhenhua) [inline website/docs/r/kubernetes_flux_configuration.html.markdown:219]:
```suggestion
* `kind` - (Required) Specifies the source kind to hold the variables to be used in substitution. Possible values are `ConfigMap` and `Secret`.
```

File: website/docs/r/kubernetes_flux_configuration.html.markdown [hunk 5]
```diff
@@ -192,6 +206,22 @@ A `git_repository` block supports the following:
...
+
+* `name` - (Required) Specifies the name of the ConfigMap/Secret that holds the variables to be used in substitution.
+
+* `optional` - (Optional) Set to True to proceed without ConfigMap/Secret, if it is not present. Defaults to `True`.
```
Comment (ms-zhenhua) [inline website/docs/r/kubernetes_flux_configuration.html.markdown:223]:
```suggestion
* `optional` - (Optional) Whether to proceed without ConfigMap/Secret if it is not present. Defaults to `false`.
```

**Review by ms-zhenhua — CHANGES_REQUESTED** [review]
Hi @jayctran, 
thank you for your updates.  There are still some minor comments about the doc.  If we can fix those up, this should be good to go 👍

**Comment by jayctran** [comment]
> Hi @jayctran, thank you for your updates. There are still some minor comments about the doc. If we can fix those up, this should be good to go 👍

thanks @ms-zhenhua, I've added those suggested commits in now.

**Review by ms-zhenhua — APPROVED** [review]
Hi @jayctran,

Thank you for your updates. LGTM~

**Comment by jayctran** [comment]
Rebased to keep branch up to date

**Comment by jayctran** [comment]
Rebased to keep branch up to date

**Review by stephybun — COMMENTED** [review]
Sorry for the long wait @jayctran and thanks for updating the tests that were using the `ARM_CLIENT_ID` environment variable.

I made a final pass and left a few minor suggestions and comments. Once those are resolved I think this should be good to go!

File: website/docs/r/kubernetes_flux_configuration.html.markdown [hunk 6]
```diff
@@ -104,6 +114,10 @@ A `kustomizations` block supports the following:
 
 * `depends_on` - (Optional) Specifies other kustomizations that this kustomization depends on. This kustomization will not reconcile until all dependencies have completed their reconciliation.
 
+* `post_build` - (Optional) A `post_build` block as defined below.
+
+* `wait` - (Optional) Whether to enable health check for all Kubernetes objects created by this Kustomization. Defaults to `true`.
```
[the remark this answers is not in the document]
Comment (jayctran) [reply website/docs/r/kubernetes_flux_configuration.html.markdown:119]: Thanks for the review @stephybun, I've left it as "wait" as it's the property used for the FluxConfiguration API as well as what's documented for Flux kustomizations: https://fluxcd.io/flux/components/kustomize/kustomizations/#wait

I figured using the official naming would enable any Flux user to understand and reference the official documentation as a reasonable approach to this.

Please let me know how you'd like this to proceed. Thanks!

[the remark this answers is not in the document]
Comment (stephybun) [reply website/docs/r/kubernetes_flux_configuration.html.markdown:119]: If this is known Flux nomenclature then I agree that it's fine to keep this as is.

**Review by stephybun — COMMENTED** [review]
@jayctran we had to revert something on main, could you perform a rebase? Once that's done I can approve and merge this 🙂

**Comment by jayctran** [comment]
> @jayctran we had to revert something on main, could you perform a rebase? Once that's done I can approve and merge this 🙂

Thanks @stephybun. No problem, that's been rebased and branch updated. Cheers!

**Review by stephybun — APPROVED** [review]
Thanks @jayctran LGTM 💯
